segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,413 @@
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+            return supplyAsync(targetHandle.getSegmentName(),
+                    () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+        this.closed.set(true);
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    @SneakyThrows(StreamSegmentNotExistsException.class)
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openRead", streamSegmentName);
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        }
+
+        LoggerHelpers.traceLeave(log, "openRead", traceId, streamSegmentName);
+        return FileSystemSegmentHandle.readHandle(streamSegmentName);
+    }
+
+    @SneakyThrows
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        } else if (Files.isWritable(path)) {
+            LoggerHelpers.traceLeave(log, "openRead", traceId);
+            return FileSystemSegmentHandle.writeHandle(streamSegmentName);
+        } else {
+            LoggerHelpers.traceLeave(log, "openWrite", traceId);
+            return syncOpenRead(streamSegmentName);
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) {
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle.getSegmentName(), offset, bufferOffset, length);
+
+        Path path = Paths.get(config.getFilesystemRoot(), handle.getSegmentName());
+
+        long fileSize = Files.size(path);
+        if (fileSize < offset) {
+            throw new IllegalArgumentException( String.format( "Reading at offset (%d) which is beyond the " +
+                    "current size of segment (%d).", offset, fileSize));
+        }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
```

Comment:
Like I said many times before, read this in a loop until you read all the bytes requested or you can't read anymore. The Channel.read() documentation explains it all.